### PR TITLE
Added a primer on CMake to docs and tweak source file search

### PR DESCRIPTION
**Pull request by mabruzzo**
I've been thinking about writing the "CMake Primer" ever since the last developer's workshop. Please let me know whether you think this is useful and if there is anything that doesn't make sense. Also let me know if you think I should add to anything else

The tweak to the source file search should makes the build system re-execute the globbing expressions whenever you execute the build-system. If the detected files change, the build system gets updated. This should reduce the frequency with which you need to start builds from scratch.

### CODE REVIEW

**Review by gregbryan — APPROVED** [review]
Thanks!  This is very useful -- minor typo corrections only.

File: doc/source/devel/cmake_primer.rst [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+   For reasons described above, steps generally needed to be taken to force the regeneration of the build system (e.g. deleting the build and starting again).
+   This should now be less of an issue.
+
+
+What is Ninja?
+~~~~~~~~~~~~~~
+
+`Ninja <https://ninja-build.org/>`_ is an alternative build-system to something like Make.
+Ninja was designed to be a smaller, less-feature-rich alternative to Make that is intended to be used with tools (like CMake) that generate its input files.
+These design goals can facillitate faster build-times in large projects.
```
Comment (jobordner) [inline doc/source/devel/cmake_primer.rst:220]:
```suggestion
These design goals can facilitate faster build-times in large projects.
```

File: doc/source/devel/cmake_primer.rst [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+This means that certain areas of the codebase receive less attention (e.g. less widely used compiler toolchains), which can lead to some issues that are more difficult to work around.
+Additionally, support staff at clusters are less likely to be familiar with SCons (so they are generally less equipped to help support users)
+
+.. _how_a_cmake_build_works:
+
+============================
+How Does a CMake Build Work?
+============================
+
+For people mostly familar with build systems like makefiles, CMake may seem a little unintuitive.
```
Comment (jobordner) [inline doc/source/devel/cmake_primer.rst:33]:
```suggestion
For people mostly familiar with build systems like makefiles, CMake may seem a little unintuitive.
```

File: doc/source/devel/cmake_primer.rst [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+This primer was written to provide some information at a level between the :ref:`getting_started` tutorial and that documentation.
+
+As an aside, when searching for online CMake documentation and guides, you should generally make sure that the guide was written for version 3.0 or later (sometimes called "Modern CMake").
+
+
+================================
+History of Enzo-E's build system
+================================
+
+Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. Howver, inn 2022 Enzo-E migrated to the CMake build system.
```
Comment (jobordner) [inline doc/source/devel/cmake_primer.rst]:
```suggestion
Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. However, in 2022 Enzo-E migrated to the CMake build system.
```

File: doc/source/devel/cmake_primer.rst [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+   This step involves the configuration of your build (more details about various configuration options for Enzo-E are provided in :ref:`how-to_configure_build`).
+   It's important to understand that CMake does not generally compile any code [#f1]_.
+   Instead, this step generates build-files (stored within the build-directory) that are natively understood by the user's platform.
+   For most Enzo-E user/developers (on unix-like systems), CMake will generate makefiles or (optionally) ninja files.
+   When you invoke cmake on the command-line you typically include variables related to the configuration and the final argument specifies the path to the root of the Enzo-E directory.
+   If you placed your build file at the root of your Enzo-E directory, this step might look something like:
+
+..  code-block:: bash
+
+  cmake -D-DCHARM_ROOT=<PATH/TO/charm/build-dir> \
```
Comment (jobordner) [inline doc/source/devel/cmake_primer.rst:58]:
```suggestion
  cmake -DCHARM_ROOT=<PATH/TO/charm/build-dir> \
```

File: doc/source/devel/cmake_primer.rst [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+
+What is Ninja?
+~~~~~~~~~~~~~~
+
+`Ninja <https://ninja-build.org/>`_ is an alternative build-system to something like Make.
+Ninja was designed to be a smaller, less-feature-rich alternative to Make that is intended to be used with tools (like CMake) that generate its input files.
+These design goals can facillitate faster build-times in large projects.
+
+To use ``Ninja``, you need to make sure it is installed on your machine and you need to specify ``-GNinja`` as one of the configuration arguments in the second step described in :ref:`how_a_cmake_build_works`.
```
Comment (jobordner) [inline doc/source/devel/cmake_primer.rst:222]:
```suggestion
To use ``Ninja``, you need to make sure it is installed on your machine and you need to specify ``-G Ninja`` as one of the configuration arguments in the second step described in :ref:`how_a_cmake_build_works`.
```

**Review by jobordner — APPROVED** [review]

File: doc/source/devel/cmake_primer.rst (hunk 3 above)
Comment (mabruzzo) [inline doc/source/devel/cmake_primer.rst]:
```suggestion
Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. However, inn 2022 Enzo-E migrated to the CMake build system.
```

**Pull request merged**